fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/ScalarApply2JoinRule.java

```diff
@@ -48,32 +64,139 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
-            output.put(apply.getOutput(), apply.getSubqueryOperator());
+        // count aggregate
+        Map<ColumnRefOperator, CallOperator> aggregates = Maps.newHashMap();
+        CallOperator countRowsCallOp = SubqueryUtils.createCountRowsOperator();
+        ColumnRefOperator countRows =
+                factory.create("countRows", countRowsCallOp.getType(), countRowsCallOp.isNullable());
+        aggregates.put(countRows, countRowsCallOp);
+
+        // any_value aggregate
+        ScalarOperator subqueryOperator = apply.getSubqueryOperator();
+        Function anyValueFn = Expr.getBuiltinFunction(FunctionSet.ANY_VALUE, new Type[] {subqueryOperator.getType()},
+                Function.CompareMode.IS_IDENTICAL);
+        CallOperator anyValueCallOp = new CallOperator(FunctionSet.ANY_VALUE, subqueryOperator.getType(),
+                Lists.newArrayList(subqueryOperator), anyValueFn);
+        ColumnRefOperator anyValue = factory.create("anyValue", anyValueCallOp.getType(), anyValueCallOp.isNullable());
+        aggregates.put(anyValue, anyValueCallOp);
+
+        OptExpression newAggOpt = OptExpression.create(
+                new LogicalAggregationOperator(AggType.GLOBAL, countAggregateGroupBys, aggregates),
+                input.inputAt(1));
+
+        /*
+         * Step2: build left outer join
+         * t0 LEFT OUTER JOIN t1a
+         * ON t0.v1 = t1a.v1
+         *
+         * If there exist subquery predicate, i.e. t0.v2 > ${subquery_output}
+         * we should put assert predicate above the subquery predicate to avoid filtering
+         * rows that shouldn't be filtered.
+         * But it is guaranteed that the subquery predicate will be push down
+         * to the left outer join operator, so we can safely put assert predicate right here.
+         */
+        IsNullPredicateOperator countRowsIsNullPredicate = new IsNullPredicateOperator(countRows);
+        BinaryPredicateOperator countRowsLEOneRowPredicate =
+                new BinaryPredicateOperator(BinaryPredicateOperator.BinaryType.LE, countRows,
+                        ConstantOperator.createBigint(1));
+        PredicateOperator countRowsPredicate =
+                new CompoundPredicateOperator(CompoundPredicateOperator.CompoundType.OR, countRowsIsNullPredicate,
+                        countRowsLEOneRowPredicate);
+        Function assertTrueFn = Expr.getBuiltinFunction(FunctionSet.ASSERT_TRUE, new Type[] {Type.BOOLEAN},
+                Function.CompareMode.IS_IDENTICAL);
+        CallOperator assertTrueCallOp =
+                new CallOperator(FunctionSet.ASSERT_TRUE, Type.BOOLEAN,
+                        Collections.singletonList(countRowsPredicate),
+                        assertTrueFn);
+        LogicalJoinOperator joinOp = new LogicalJoinOperator.Builder()
+                .setJoinType(JoinOperator.LEFT_OUTER_JOIN)
+                .setOnPredicate(Utils.compoundAnd(correlationPredicatePair.first))
+                .setPredicate(assertTrueCallOp)
```

Comment:
predicate t0.v2 > t1a.anyValue do not used in join operator? where this predicate used?

Reply:
This predicate is not belongs to this rule pattern, we just add a project operator to map `subquery output` to `anyValue`.